util/compression_test.cc

```diff
@@ -0,0 +1,179 @@
+    options.statistics = ROCKSDB_NAMESPACE::CreateDBStatistics();
+    options.statistics->set_stats_level(StatsLevel::kExceptTimeForMutex);
+    BlockBasedTableOptions bbto;
+    bbto.enable_index_compression = false;
+    options.table_factory.reset(NewBlockBasedTableFactory(bbto));
+    DestroyAndReopen(options);
+  }
+  uint64_t PopStat(Tickers t) {
+    return options.statistics->getAndResetTickerCount(t);
+  }
+  bool CompressionFriendlyPut(const int no_of_kvs, const int size_of_value) {
+    auto value = std::string(size_of_value, 'A');
+    for (int i = 0; i < no_of_kvs; ++i) {
+      auto status = Put(Key(key_cursor_), value);
+      EXPECT_EQ(status.ok(), true);
+      key_cursor_++;
+    }
+    return true;
+  }
+  bool CompressionUnfriendlyPut(const int no_of_kvs, const int size_of_value) {
+    auto value = rnd_.RandomBinaryString(size_of_value);
+    for (int i = 0; i < no_of_kvs; ++i) {
+      auto status = Put(Key(key_cursor_), value);
+      EXPECT_EQ(status.ok(), true);
+      key_cursor_++;
+    }
+    return true;
+  }
+};
+
+// test case just to make sure auto compression manager is working
+TEST_F(DBAutoSkip, AutoSkipCompressionManager) {
+  if (GetSupportedCompressions().size() > 1) {
+    // AutoSkipCompressionManager starts with rejection ratio 0 i.e. compression
+    // enabled
+    constexpr int kCount =
+        5000;  // high enough for compression manager to register the changes
+    // enough data to change the decision
+    const int kValueSize = 20000;
+    auto set_exploration = [&](void* arg) {
+      bool* exploration = static_cast<bool*>(arg);
+      *exploration = true;
+    };
+    SyncPoint::GetInstance()->DisableProcessing();
+    SyncPoint::GetInstance()->ClearAllCallBacks();
+    SyncPoint::GetInstance()->SetCallBack(
+        "AutoSkipCompressorWrapper::CompressBlock::exploitOrExplore",
+        set_exploration);
+    SyncPoint::GetInstance()->EnableProcessing();
+    CompressionUnfriendlyPut(kCount, kValueSize);
+    ASSERT_OK(Flush());
+    SyncPoint::GetInstance()->DisableProcessing();
+    // reset the following stats to zero
+    PopStat(NUMBER_BLOCK_COMPRESSED);
+    PopStat(NUMBER_BLOCK_COMPRESSION_BYPASSED);
+    PopStat(NUMBER_BLOCK_COMPRESSION_REJECTED);
+    // should stick with the not compressing decision
+    CompressionUnfriendlyPut(kCount, kValueSize);
+    ASSERT_OK(Flush());
+    // Test the compression is disabled
+    // Make sure that Compressor output is properly calculated as bypassed
+    auto compressed_count = PopStat(NUMBER_BLOCK_COMPRESSED);
+    auto bypassed_count = PopStat(NUMBER_BLOCK_COMPRESSION_BYPASSED);
+    auto rejected_count = PopStat(NUMBER_BLOCK_COMPRESSION_REJECTED);
+    auto rejection_ratio =
+        rejected_count * 100 / (compressed_count + rejected_count);
+    EXPECT_GT(rejection_ratio, 50);
+    auto bypassed_rate = bypassed_count * 100 /
+                         (compressed_count + rejected_count + bypassed_count);
+    EXPECT_GT(bypassed_rate, 50);
```

Comment:
These verification is still confusing and I might miss something. 

For `kCount` incompressible key values, isn't it true that bypass_ratio = 1 - exploration percentage (TEST_GetExplorationPercentage()) and rejection_ratio = 1 -  bypass_ratio = exploration percentage. So for all incompressible key values, they either get explored (which can be observed by rejection ratio) or obey the prediction (which can be observed by bypass ratio).